### PR TITLE
feat(Core/Packets): rework HandleWhoOpcode

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Enable old code for who list cache
- Make safely code for `PROCESS_THREADSAFE`
- Use cache for avoids usage of hashmapholder mutex
- Add new metric timer `world_update_time` `Update who list`

## Issues Addressed:

- Closes 

## SOURCE:
- https://github.com/TrinityCore/TrinityCore/commit/88f746935ff6a9055c5712997254aff63b406352

## Tests Performed:
- `acc level 4`
image
- `acc level 0`
image

## How to Test the Changes:
1. Check who

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Cache/WhoListCacheMgr.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        if (!Utf8toWStr(playerName, widePlayerName))
+            continue;
+
+        wstrToLower(widePlayerName);
+
+        std::string guildName = sGuildMgr->GetGuildNameById(player->GetGuildId());
+        std::wstring wideGuildName;
+
+        if (!Utf8toWStr(guildName, wideGuildName))
+            continue;
```
Comment (Nefertumm) [inline src/server/game/Cache/WhoListCacheMgr.cpp:54]: brackets

File: src/server/game/Handlers/MiscHandler.cpp [hunk 2]
```diff
@@ -211,197 +212,206 @@ void WorldSession::HandleWhoOpcode(WorldPacket& recvData)
...
-    // client send in case not set max level value 100 but Trinity supports 255 max level,
+    wstrToLower(wpacketPlayerName);
+    wstrToLower(wpacketGuildName);;
+
+    // client send in case not set max level value 100 but Acore supports 255 max level,
     // update it to show GMs with characters after 100 level
-    if (level_max >= MAX_LEVEL)
-        level_max = STRONG_MAX_LEVEL;
+    if (levelMax >= MAX_LEVEL)
+        levelMax = STRONG_MAX_LEVEL;
```
Comment (Nefertumm) [inline src/server/game/Handlers/MiscHandler.cpp:275]: same

File: src/server/game/Handlers/MiscHandler.cpp [hunk 3]
```diff
@@ -211,197 +212,206 @@ void WorldSession::HandleWhoOpcode(WorldPacket& recvData)
...
-        LOG_DEBUG("network", "String %u: %s", i, temp.c_str());
+        FMT_LOG_DEBUG("network.who", "String {}: {}", i, temp);
     }
 
-    std::wstring wplayer_name;
-    std::wstring wguild_name;
-    if (!(Utf8toWStr(player_name, wplayer_name) && Utf8toWStr(guild_name, wguild_name)))
+    std::wstring wpacketPlayerName;
+    std::wstring wpacketGuildName;
+    if (!(Utf8toWStr(packetPlayerName, wpacketPlayerName) && Utf8toWStr(packetGuildName, wpacketGuildName)))
```
Comment (Nefertumm) [inline src/server/game/Handlers/MiscHandler.cpp:266]: also brackets after this if

File: src/server/game/Handlers/MiscHandler.cpp [hunk 4]
```diff
@@ -211,197 +212,206 @@ void WorldSession::HandleWhoOpcode(WorldPacket& recvData)
...
+    recvData >> racemask;                                   // race mask
+    recvData >> classmask;                                  // class mask
+    recvData >> zonesCount;                                 // zones count, client limit = 10 (2.0.10)
 
-    recvData >> racemask;                                  // race mask
-    recvData >> classmask;                                 // class mask
-    recvData >> zones_count;                               // zones count, client limit = 10 (2.0.10)
-
-    if (zones_count > 10)
+    if (zonesCount > 10)
```
Comment (Nefertumm) [inline src/server/game/Handlers/MiscHandler.cpp:231]: same

File: src/server/game/Handlers/MiscHandler.cpp [hunk 5]
```diff
@@ -211,197 +212,206 @@ void WorldSession::HandleWhoOpcode(WorldPacket& recvData)
...
         zoneids[i] = temp;
-        LOG_DEBUG("network", "Zone %u: %u", i, zoneids[i]);
+        FMT_LOG_DEBUG("network.who", "Zone {}: {}", i, zoneids[i]);
     }
 
-    recvData >> str_count;                                 // user entered strings count, client limit=4 (checked on 2.0.10)
+    recvData >> strCount;                                   // user entered strings count, client limit=4 (checked on 2.0.10)
 
-    if (str_count > 4)
+    if (strCount > 4)
```
Comment (Nefertumm) [inline src/server/game/Handlers/MiscHandler.cpp:244]: same

**Review by Nefertumm — APPROVED** [review]

File: src/server/game/Cache/WhoListCacheMgr.h [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+//    uint8 gender;
+//    std::wstring wpname;
+//    std::wstring wgname;
+//    std::string aname;
+//    std::string pname;
+//    std::string gname;
+//
+//    WhoListPlayerInfo(TeamId teamId, AccountTypes security, uint8 level, uint8 clas, uint8 race, uint32 zoneid, uint8 gender, std::wstring wpname, std::wstring wgname, std::string aname, std::string pname, std::string gname) :
+//        teamId(teamId), security(security), level(level), clas(clas), race(race), zoneid(zoneid), gender(gender), wpname(wpname), wgname(wgname), aname(aname), pname(pname), gname(gname) {}
+//};
```
Comment (Winfidonarleyan) [inline src/server/game/Cache/WhoListCacheMgr.h:42]:
```suggestion
```

**Comment by MysterioPRM** [comment]
PR tested .Works good


GM acc:
image

Players acc:
image

**Comment by Winfidonarleyan** [comment]
> PR tested .Works good Source compiled without any modules
> 
> GM acc: image
> 
> Players acc: image

Thanks for test

**Pull request merged**